net/socks5/conn.go

```diff
@@ -0,0 +1,151 @@
+package socks5
+
+import (
+	"context"
+	"fmt"
+	"io"
+	"log"
+	"net"
+	"strconv"
+)
+
+type DialContext func(ctx context.Context, network, addr string) (net.Conn, error)
+
+// Conn represents a SOCKS5 connection for client to reach
+// server. The struct is filled by each of the internal methods
+// in turn as the transaction progresses.
+type Conn struct {
+	dialContext DialContext
+	client      net.Conn
+	server      net.Conn
+	methods     []AuthMethod
+	request     *Request
+}
+
+// NewConn creates a new SOCKS5 connection that uses the default
+// dialing context to talk to the SOCKS5 server
+func NewConn(clientConn net.Conn) *Conn {
+	dialer := &net.Dialer{}
+	return NewConnWithDialContext(clientConn, dialer.DialContext)
+}
+
+// NewConnWithDialContext creates a new SOCKS5 connection that uses
+// a custom dialing context to talk to the SOCKS5 server
+func NewConnWithDialContext(clientConn net.Conn, dialContext DialContext) *Conn {
+	return &Conn{
+		client:      clientConn,
+		dialContext: dialContext,
+	}
+}
+
+func (conn *Conn) init() error {
+	buf := make([]byte, MaxInitRequestSize)
+	n, err := conn.client.Read(buf)
```

Comment:
You probably want to use io.ReadFull here or one of the other package io helpers; io.Reader can hand bytes to you one at a time (or none at a time).

Reply:
io.ReadFull reads exactly len(buf) bytes of data, but I don't think we'll reach that packet size in the normal case. Would it be a good idea to use io.ReadAll, which reads until EOF?

Reply:
ReadAtLeast then: https://golang.org/pkg/io/#ReadAtLeast